### ISSUE TITLE
Where is the torch / flashlight button?

### ISSUE THREAD

**Issue by mihaelmilea** [ubports/ubports-touch#223]
- Device: Turbo
- Channel: Devel
- Build: 20

## Steps to reproduce

1. I pull down the top settings menu
2. I search everywhere and in all sections

## Expected behavoir

I understood there would be a torch button somewhere.

## Actual behavior

I cannot find it. Is it implemented?

## Logfiles and additional information

...

**Comment by Flohack74** [comment]
Hmm on hammerhead it is there... Someone with a turbo needs to confirm. But also @rubencarneiro mentioned it is not always visible. Maybe we got a problem with loading it correctly?

**Comment by mihaelmilea** [comment]
@Stefano0101 - can you confirm this please?

**Comment by Stefano0101** [comment]
there is nfs-prodriver-indicato-torch icon, but not the actual feature, just a blank black space nothing else

**Comment by mihaelmilea** [comment]
Yes, there is the indicator-torch in the open store and that installs the indicator in the upper menu that becomes visible after a restart and there I can find on and off menus but those don't do anything. Is this the new torch feature or is it something else somewhere in the battery menu?

**Comment by Stefano0101** [comment]
Yeah, I know, tried that indicator, did not worked, uninstalled it and this one is there by default. I have not installed this one. Anyway, it doesn't work, as there is nothing to click/tap on.28.08.2017, 12:50, "Mihael Milea" <notifications@github.com>:Yes, there is the indicator-torch in the open store and that installs the indicator in the upper menu that becomes visible after a restart and there I can find on and off menus but those don't do anything. Is this the new torch feature or is it something else somewhere in the battery menu?

**Comment by mihaelmilea** [comment]
There is nothing in my menu! Is there a way you can attach a screen shot?

**Comment by Stefano0101** [comment]
It was the residual icon, now properly uninstalled(the indicator light)  and there is noting left.

**Comment by NeoTheThird** [comment]
The Pro 5 (and some other non qcom devices) does not have open source drivers for the flashlight, so it does not appear on them and only works on Nexus 5, Oneplus One and Fairphone 2, iirc. The only possible way to fix this would be using the camera API (or keep digging for the mediatek flashlight drivers).

**Comment by NeoTheThird** [comment]
@mihaelmilea I took the liberty to modify the description. Hope you don't mind.

**Comment by mihaelmilea** [comment]
Not at all, not at all! Thank you!

**Comment by Stefano0101** [comment]
Pro 5 has a Samsung Exynos processor.

**Comment by tomoqv** [comment]
This thread is somewhat confusing. @NeoTheThird says that flashlight works on N5, OPO and FP2, but flashlight has never worked on my OPO running 15.04 devel.

**Comment by ChloeWolfieGirl** [comment]
I beleive this is fixed next ota.

I have some features from the next Ota such as backgrounds, the channel switcher etc.

One update I have gives me the option to use a torch on the opo under the battery indicator.

So my oneplus one's torch works.

I dont want to close this incase it doesnt work on every device but we can remove bacon.

**Comment by tomoqv** [comment]
@ChloeWolfieGirl Yes, I noticed the flashlight option under battery indicator, thanks.

**Comment by NeoTheThird** [comment]
@tomoqv 
> Edit: I just discovered the Flashlight option under the Battery indicator and it works! However, uTorch and Indicator Torch from the OpenStore still don't work.

My guess is that uTorch and Indicator Torch use the camera API to activate the led which does not work currently on the Oneplus One... :( We wanted to avoid using it, since it uses more battery, but @mariogrip what do you think, could we use it for mediatek devices?

@ChloeWolfieGirl thanks for the hint.

**Comment by NeoTheThird** [comment]
In case someone is interested in adding the camera api as a fallback in case no driver is available, the implementation of the flashlight switch is done in these commits:
- https://github.com/ubports/indicator-power/commit/3f4339876ddb586b75a73fab51a54b1d4d54fdd1
- https://github.com/ubports/indicator-power/commit/efef6f25f378021c99b3078958482b3c4ffc43a2

As a reference, [indicator-torch](https://github.com/nfsprodriver/ut-indicator-torch) and [uTorch](https://code.launchpad.net/utorch) are available.

**Comment by rubencarneiro** [comment]
https://github.com/rubencarneiro/indicator-power/commit/f28b98c975008a119a10b75b17785564bee6c9aa

**Comment by mihaelmilea** [comment]
Is this working on 15.4? :)

**Comment by rubencarneiro** [comment]
it enables torch function in power indicator but cant seem to enable flash

**Comment by mihaelmilea** [comment]
How can I get it? :)

**Comment by rubencarneiro** [comment]
i didnt merge for now , until is wotking as it should so you cant get it for now

**Comment by mihaelmilea** [comment]
Ok, thanks for all your work! We all appreciate it!

**Comment by rubencarneiro** [comment]
if you can help debug, or give more info, is better, this is on MX4
![screenshot20180314_000933721](https://user-images.githubusercontent.com/6282290/37877908-42c9c754-3059-11e8-9d4b-10214cccd29c.png)

**Comment by rubencarneiro** [comment]
Since syspath is not the same to all devices, we need feeadback form devices to make it work.
